aoc2022/src/main/scala/day09.scala

```diff
@@ -0,0 +1,80 @@
+import aoc2022.FlatGrid
+import scala.io.*
+import math.*
+import scala.annotation.tailrec
+import aoc2022.Grid2D.Point
+
+
+/**
+ * PART 01:
+ *
+ * I changed my solution for part01 based on the requirements for part02. The main thing I found out was that you 
+ * can make a knot move like the previous not by just looking at the location of that previous knot: 
+ *
+ * val dist: Pointa = head - tail
+ * if math.abs(dist.x) <= 1 && math.abs(dist.y) <= 1 then don't move
+ * else tail + Point(dist.x.sign, dist.y.sign) 
+ *
+ * Once figured out, the remainder could be solved by a foldLeft, and a scanLeft on every knot in the knotList. 
+ *
+ * PART 02:
+ *
+ * After adjusting part01 to the requirements for part02, the solution was a simple run for the 9th knot. 
+ *
+ */
+
+
+object day09 extends App:
+
+  private val day: String =
+    this.getClass.getName.drop(3).init
+
+  private val start1: Long =
+    System.currentTimeMillis
+
+  private val input: List[(Char, Int)] =
+
+    def parseMove(s: String): (Char, Int) =
+      val parts = s.split(" ")
+      (parts.head.head, parts.last.toInt)
+
+    Source
+      .fromResource(s"day$day.txt")
+      .getLines
+      .toList
+      .map(parseMove)
+
+  extension(p: Point)
+    def snapToPoint(that: Point): Point =
+      val dist: Point = that - p
+      if math.abs(dist.x) <= 1 && math.abs(dist.y) <= 1 then p  // points are too close no movement needed
+      else p + Point(dist.x.sign, dist.y.sign)
+
+  case class RopeMovement(knots: List[Point], seen: Set[Point]):
+    def moveRope(dir: Point): RopeMovement =
+      val newT: List[Point] = knots.drop(1).scanLeft(knots.head + dir)((thisH: Point, thisT: Point) => thisT.snapToPoint(thisH))
+      RopeMovement(newT, seen + newT.last)
```

Comment:
`.last` is quite expensive on a `List`; although you are traversing the list above as well, consider using something else like a `Vector`

Reply:
Ah check. So if I understand correctly, when you're just taking/appending from the `head` of a List, the List is a more suitable data structure than a Vector. However, when you're also reading values from an index `i` (as is the case with the `.last` method), then a Vector is more suitable?